README.md
<a href="https://chrome.google.com/webstore/detail/chatgpt-history/jjdnakkfjnnbbckhifcfchagnpofjffo/"><img src="https://user-images.githubusercontent.com/12115686/206926802-0461dc64-84cd-42de-8c17-74a7ee64528c.png" style="width: 180px !important; height: 50px !important"></a> <a href="https://addons.mozilla.org/en-US/firefox/addon/chatgpt-history/"><img src="https://user-images.githubusercontent.com/12115686/207746497-4b4ba50c-c579-42ad-b2e9-9164073499db.png" style="width: 180 !important; height: 50px !important"></a> <a href="https://www.reddit.com/r/ChatGPTPromptGenius/"><img src="https://user-images.githubusercontent.com/12115686/211184170-6aea6981-abd4-447c-bd3d-199d1688011f.png" style="width: 50px !important"></a> <a href="https://ko-fi.com/bennyfi" target="_blank"><img src="https://storage.ko-fi.com/cdn/kofi3.png?v=3" alt="Buy Me A Coffee" style="height: 40px !important;width: 173px !important;" ></a> <a href="http://creativecommons.org/licenses/by-nc-sa/4.0/"><img src="https://licensebuttons.net/l/by-nc-sa/4.0/88x31.png" style="height: 48px !important"></a>

<a href="https://chrome.google.com/webstore/detail/chatgpt-history/jjdnakkfjnnbbckhifcfchagnpofjffo/"><img alt="Chrome Web Store" src="https://img.shields.io/chrome-web-store/users/jjdnakkfjnnbbckhifcfchagnpofjffo?color=red&label=Chrome%20Users"></a>
<a href="https://addons.mozilla.org/en-US/firefox/addon/chatgpt-history/"><img alt="Mozilla Add-on" src="https://img.shields.io/amo/users/chatgpt-history?label=Firefox%20Users"></a> ![Chrome Web Store](https://img.shields.io/chrome-web-store/stars/jjdnakkfjnnbbckhifcfchagnpofjffo?color=yellow) [![Crowdin](https://badges.crowdin.net/chatgpt-prompt-genius/localized.svg)](https://crowdin.com/project/chatgpt-prompt-genius)

[Advertise with us](https://github.com/benf2004/ChatGPT-Prompt-Genius/blob/master/ADVERTISING.md)

[**Help translate**](https://crowdin.com/project/chatgpt-prompt-genius/invite/public?h=6b3278f0cfb94d5a92e0c504dd273f7a1645597&show-join)


# ChatGPT Prompt Genius  
<em>Written by ChatGPT, of course</em>

🎉 Welcome to ChatGPT Prompt Genius, a free, open-source browser extension that helps you 🔍 discover, share, import, and use the best prompts for ChatGPT. You can also 💾 save your chat history locally so you can easily review past conversations and refer to them at a later time.

You can use the extension's prompt templates feature to easily find and add prompts to your collection. You can search, categorize, and select prompts right on the page, making it easy to find creative and productive ways to use ChatGPT.

🎨Add themes like SMS, cozy fireplace, and hacker on the ChatGPT page. 

To use the history-saving feature, simply open ChatGPT and start chatting 💬 as you normally would. The extension will automatically save your conversation history in your browser. You can access your saved history by clicking on the extension icon 🔍 in the top right corner of your browser. The extension rerenders your conversation in a style that closely matches ChatGPT, including code rendering & copying.

On the explore page, you can 🔖 bookmark threads or 🔍 search through your threads & prompt templates. You can also save your chats as html/pdf/png 📄 right on ChatGPT.

We hope you find ChatGPT Prompt Genius useful and enjoy using it to find and save your ChatGPT prompts. If you have any feedback or suggestions for improvement, please don't hesitate to fill out the feedback form. Thank you for using ChatGPT Prompt Genius!

Prompt Creator:

<img width="640" alt="Screen Shot 2023-01-07 at 11 34 16 PM" src="https://user-images.githubusercontent.com/12115686/211184017-57b816d8-020b-4cbf-a67b-579b142068f3.png">

Curated Prompt View:

<img width="640" alt="Screen Shot 2023-01-07 at 11 35 09 PM" src="https://user-images.githubusercontent.com/12115686/211184023-777ff59d-a20d-4510-a7ae-30d9d86682ba.png">

SMS Theme:

<img width="640" alt="SMS" src="https://user-images.githubusercontent.com/12115686/211184025-b515a3e8-33f3-4e4a-aae8-ec1a30af0a50.png">


Thread Explorer:
![screely-1670886428256](https://user-images.githubusercontent.com/12115686/207233691-92e31001-c045-4f77-bd3f-bc9170814360.png)



Thread View:
![screely-1670886518332](https://user-images.githubusercontent.com/12115686/207233426-e932fe34-0ffe-45c4-9f45-7a098e062f50.png)


## TODO
Please see TODO.md and the issues tab. These enhancements would make the platform even more user-friendly and useful for our users. Thank you for considering contributing to the project!

## Installation
- Chrome - Install from the <a href="https://chrome.google.com/webstore/detail/chatgpt-history/jjdnakkfjnnbbckhifcfchagnpofjffo/">Chrome Web Store</a>
- Firefox - Install from <a href="https://addons.mozilla.org/en-US/firefox/addon/chatgpt-history/">Fire Fox Add-ons library</a>
- Run locally - clone the repo and run the appropriate build script for your OS. Use mv2 for Firefox and mv3 for Chrome.

## Structure
<em>It's all vanilla, baby</em>
src/content-scripts: includes all content scripts (including history scraper, prompt injection, export buttons, and reddit stuff)
src/external-js: external libraries 
src/icons: extension icons
src/pages: includes all the extension pages (HTML + CSS + JS) for viewing thread history & prompt templates
src/themes: CSS themes for styling ChatGPT
background.js: listens for user clicks and handles message passing
utility.js: common JS

CSS: Bootstrap & Font Awesome Icons + some custom  



## Credit 
- Thanks to @liady for the libraries used for the export functions. See <a href="https://github.com/liady/ChatGPT-pdf">ChatGPT-pdf</a> for more.
- Thanks to <a href="https://github.com/mohalobaidi/EnhancedChatGPT">Enhanced ChatGPT's</a> code for in-page prompt insertion.
- Many thanks to @aivct for contributing many new features.

## License
Shield: [![CC BY-NC-SA 4.0][cc-by-nc-sa-shield]][cc-by-nc-sa] 

This work is licensed under a
[Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License][cc-by-nc-sa].

[![CC BY-NC-SA 4.0][cc-by-nc-sa-image]][cc-by-nc-sa]

[cc-by-nc-sa]: http://creativecommons.org/licenses/by-nc-sa/4.0/ 
[cc-by-nc-sa-image]: https://licensebuttons.net/l/by-nc-sa/4.0/88x31.png 
[cc-by-nc-sa-shield]: https://img.shields.io/badge/License-CC%20BY--NC--SA%204.0-lightgrey.svg 

Proper Attribution (include in the prominent file):  

```
/* ChatGPT Prompt Genius - https://github.com/benf2004/ChatGPT-Prompt-Genius
@benf2004 - https://github.com/benf2004/ - Creator
@aivct - https://github.com/aivct/ - Major Contributor
View the full list of contributors at https://github.com/benf2004/ChatGPT-Prompt-Genius/graphs/contributors
*/
```



CHANGELOG.md
### 3.2.9 -November 9, 2023

- Stop injecting prompts while I finish v4

### 3.2.8 - October 24, 2023

- Fix scrolling issue

### 3.2.7 - October 2, 2023

- Include JSpdf object locally
- Fix extension warning

### 3.2.6 - September 15, 2023

- Rename external-js to included-libraries
- Fix button bug
- Other minor updates

### 3.2.5 - September 9, 2023

- Add option to export prompts to CSV
- Fix highlighting issue
- Improve zh_TW translations

### 3.2.4 - September 1, 2023

- Fix #322

### 3.2.3 - August 21, 2023

- Fix text div detection

### 3.2.2  - August 20, 2023

- Fix #318, #306, & #308 (error cloud syncing when no prompts)

### 3.2.1 - August 14, 2023 

- Fixed regenerate bug being removed

### 3.2.0 - August 14, 2023

- Fixed how uninstall URL is delivered
- Bug fixes

### 3.1.9 - August 11, 2023

- Fixes on the fixes
- Prompt hotkey should work now

### 3.1.8 - August 10, 2023

- Fix styling (cards, search/category, variable modal)
- Fix "Use This prompt button"
- Remove new suggested prompts from input

### 3.1.7 - August 8, 2023

- Localize ads to user region, fallback to language, fallback to US

### 3.1.6 - August 3, 2023

- Localize ads to users region

### v3.1.5 - July 29, 2023

- Turkish & Italian translations

### v3.1.4 - July 28, 2023

- Localized export to markdown
- Switched export to markdown to Turndown (even though showdown is supposed to be bidirectional, it doesn't really work)
- Markdown syntax is consistent, including support for MathJax
- Removed ShareGPT (sherlocked)
- Fixed icon in thread
- Fixed GPT-4 overlays
- Moved descriptions from extension source to outside

### v3.1.3 - July 19, 2023

- Switch to ko-fi instead of buymeacoffee

### v3.1.2 - July 14, 2023

- Improved zh_TW translations

### v3.1.1 - June 26, 2023

- Rename extension AI prompt genius per legal complaint

### v3.1.0 - June 17, 2023

- Fixes #300

### v3.0.9 - June 16, 2023 

- Updated variable modal to support multiple lines
- Fixed bug where sidebar would not appear
- Removed navbar arrow (sherlocked)

### v3.0.8 - May 23, 2023

- Fixed export bug

### v3.0.7 - May 18, 2023

- Fixed bug where settings wouldn't save

### v3.0.6 - May 16, 2023

- New icons 
- free, legit cdn

### v3.0.5

- Removed Reddit Scripts
- Fixed SMS theme
- Added oninstall page

### v3.0.4 

- Fix bug #272/#279/#283

### v3.0.3 - May 8, 2023

- German language
- Tutorial link

### v3.0.2 - May 5, 2023

- fix bug #274

### v3.0.1 - May 5, 2023

- Theme select bug
- Theme fixes
- Remove clipboard copy (now native)
- New link for "What's New?"

### v3.0.0 - April 24, 2023

- See the [what's new page](https://docs.aipromptgenius.app/what-s-new/)

### v2.5.6 - April 22, 2023

- Fix login bug

### v2.5.5 - April 18, 2023

- Remove blank translate.js file
- Fix scraper
- Plus network input bug (hide it)

### v2.5.4 - April 17, 2023

- Fix URL changes for real

### v2.5.3 - April 14, 2023

- Hotfix to fix URL changes

### v2.5.2 - April 11, 2023

- Add identity permission for cloud syncing
- Remove advanced theme editor

### v2.5.1 - April 7, 2023

- Cloud syncing - Chrome syncing
- Light/dark mode bug
- Tooltips bug
- Theme editor bug
- Writing multiple new prompts bug

### v2.5.0 - April 3, 2023

- Russian language

### v2.4.8 & 2.4.9 - April 3, 2023

- even HOTTER fixes

### v2.4.7 - April 3, 2023

- Move resync page
- Add abort controller
- New thing

### v2.4.6 - March 28, 2023

- Update ad placement on prompts.html
- Fix tag issue
- Made translation more efficient 
- Drag prompts to reorder

### v2.4.5 - March 27, 2023

- French
- Update interval is 10s

### v2.4.4 - March 25, 2023

- Ukrainian, Traditional Chinese
- Localized CWS description
- Fixed manifest for localization

### v2.4.3 - March 22, 2023

- Missing i18n tag
- Fixed bookmarks?
- Load threads in order

### v2.4.2 - March 22, 2023 

- Fixed replacing left tags in code blocks
- Localization
- Fixed double scrollbar issue

### v2.4.1 - March 20, 2023

- History v2: sync history based off of network storage
- Add ad to prompts home page
- Replace < with "& lt;"
- Removed options UI from manifest

### v2.4.0 - March 20, 2023

- Fixed share link bug for ChatGPT Plus
- Users see ad on uninstall

### v2.3.9 - March 16, 2023

- Bug fixes for reinjecting prompt and detecting if the user is plus 

### v2.3.8 - March 12, 2023

- Added naming the code blocks to the history scraper

### v2.3.7 - March 11, 2023

- Manually identify as a ChatGPT Plus user as a fallback
- Option to disable in-page prompt injection per user request

### v2.3.6 - March 9, 2023

- Build prompts instead of replacing

### v2.3.5 - March 8, 2023

- Fixed delete history

### v2.3.4 - March 8, 2023

- Don't get cached ad
- Removed toast
- Changed first demo prompt
- Fixed ads for firefox

### v2.3.3 - March 7, 2023

- Fixed issue where ad would not re-inject after clicking "new chat" 

### v2.3.2 - March 4, 2023

- Fixed trailing space thing with tags

### v2.3.1 - March 4, 2023

- Fixed filtering tags again

### v2.3.0 - March 4, 2023

- Fixed re-adding tooltips after filtering/searching
- Added in page tags
- Minor code cleanup on prompt injection

### v2.2.9 - March 3, 2023 [hotfix]

- Fixed search function on prompts

### v2.2.8 - March 3, 2023

- Added tags to prompts

### v2.2.7 - March 2, 2023

- Added separate storage page

### v2.2.6  - March 2, 2023

- Moved position of share button items
- Added compact view
- Added links to view and discover prompts

### v2.2.5 - March 1, 2023

 - Fixed issue with re-entering prompt templates

### v2.2.4 - March 1, 2023

- Fixed typo in popup name of extension
- Added support for non-intrusive [ads](https://github.com/benf2004/ChatGPT-Prompt-Genius/blob/master/ADVERTISING.md)

### v2.2.3 - March 1, 2023

- Fixed in-page search

### v2.2.2 - February 24, 2023 [hotfix]

- Hotfix for v2.2.1

### v2.2.1 - February 23, 2023

- Added support for ChatGPT Plus

### v2.2.0 - February 18, 2023

- Fixed jumbled nav on zoomed in screen


src/thread.js
if (typeof browser === "undefined") {
    browser = chrome
}
hljs.highlightAll();
// Get the URL of the current page
const url = new URL(window.location.href);

// Get the value of the "name" parameter in the query string
const thread_id = url.searchParams.get("thread");

const h_template = document.querySelector("#human")
const b_template = document.querySelector("#bot")

let main = document.querySelector("#main");
let branch_state;
let convo;
browser.storage.local.get(['threads']).then((result) => {
    let t = result.threads
    let c = getObjectById(thread_id, t)
	convo = c.convo;
	// some of the older threads don't have a branch_state object. 
	let b = c.branch_state;
	if(!b)
	{
		console.log(`Cannot find branch state, loading convo instead...`);
		load_thread(convo);
	}
	else 
	{
		console.log(`Loading branch state...`);
		branch_state = new TreeNode();
		branch_state.fromJSON(b);
		load_branched_thread();
	}
});

function load_thread(c){
    for (let i = 0; i < c.length; i++) {
        let human = i % 2 === 0;
        let bar = `<div class="flex items-center relative text-gray-200 bg-gray-800 px-4 py-2 text-xs font-sans"><button class="flex ml-auto gap-2"><svg stroke="currentColor" fill="none" stroke-width="2" viewBox="0 0 24 24" stroke-linecap="round" stroke-linejoin="round" class="h-4 w-4" height="1em" width="1em" xmlns="http://www.w3.org/2000/svg"><path d="M16 4h2a2 2 0 0 1 2 2v14a2 2 0 0 1-2 2H6a2 2 0 0 1-2-2V6a2 2 0 0 1 2-2h2"></path><rect x="8" y="2" width="8" height="4" rx="1" ry="1"></rect></svg>Copy code</button></div>`
        if (human) {
            var temp = h_template.content.cloneNode(true);
            temp.querySelector(".text").innerHTML = `<p>${c[i]}</p>`
            main.appendChild(temp)
        }
        else{
            var temp = b_template.content.cloneNode(true);
            let clipboard = `<i class="fa-regular clipboard fa-clipboard"></i>`
            let copy_bar = `<div class="p-2 copy float-right">${clipboard} &nbsp; Copy code</div>`
            temp.querySelector(".text").innerHTML = c[i].replaceAll(bar, copy_bar).replaceAll(`<div class="p-4">`, "<div>") // fixes formatting for weird code divs
            main.appendChild(temp)
        }
    }
    copy_setup();
}

function load_branched_thread()
{
	// reset 
	main.innerHTML = "";
	
	let fake_convo = branch_state.getCurrentData();
	let current_leaf = branch_state;
	// first element is always empty as the head node
	if(fake_convo[0] === undefined)
	{
		fake_convo.shift(); 
	}
	else 
	{
		console.warn(`load_branched_thread: data is malformed, weird things are going to happen.`);
	}
	
	for (let i = 0; i < fake_convo.length; i++) {
		let human = i % 2 === 0;
        let bar = `<div class="flex items-center relative text-gray-200 bg-gray-800 px-4 py-2 text-xs font-sans"><button class="flex ml-auto gap-2"><svg stroke="currentColor" fill="none" stroke-width="2" viewBox="0 0 24 24" stroke-linecap="round" stroke-linejoin="round" class="h-4 w-4" height="1em" width="1em" xmlns="http://www.w3.org/2000/svg"><path d="M16 4h2a2 2 0 0 1 2 2v14a2 2 0 0 1-2 2H6a2 2 0 0 1-2-2V6a2 2 0 0 1 2-2h2"></path><rect x="8" y="2" width="8" height="4" rx="1" ry="1"></rect></svg>Copy code</button></div>`
		
		let current_leaf_index;
		let leaves_length = 0;
		
		let data_parent_leaf = current_leaf;
		
		if(current_leaf)
		{
			leaves_length = current_leaf.getNumberOfLeaves();
			current_leaf_index = current_leaf.getCurrentLeafIndex();
			// update for next loop iter 
			current_leaf = current_leaf.getCurrentLeaf();
		}
		// set text 
		var temp;
		if(human) {
			temp = h_template.content.cloneNode(true);
		}
		else {
			temp = b_template.content.cloneNode(true);
		}
		
		
		if(leaves_length > 1)
		{
			// temp.querySelector(".branch").innerHTML = `<div class="text-xs flex items-center justify-center gap-1 invisible absolute left-0 top-2 -ml-4 -translate-x-full group-hover:visible"><button class="dark:text-white disabled:text-gray-300 dark:disabled:text-gray-400"><svg stroke="currentColor" fill="none" stroke-width="1.5" viewBox="0 0 24 24" stroke-linecap="round" stroke-linejoin="round" class="h-3 w-3" height="1em" width="1em" xmlns="http://www.w3.org/2000/svg"><polyline points="15 18 9 12 15 6"></polyline></svg></button><span class="flex-grow flex-shrink-0">${current_leaf_index+1} / ${leaves_length}</span><button disabled="" class="dark:text-white disabled:text-gray-300 dark:disabled:text-gray-400"><svg stroke="currentColor" fill="none" stroke-width="1.5" viewBox="0 0 24 24" stroke-linecap="round" stroke-linejoin="round" class="h-3 w-3" height="1em" width="1em" xmlns="http://www.w3.org/2000/svg"><polyline points="9 18 15 12 9 6"></polyline></svg></button></div>`;
			
			let branchSelectorElement = temp.querySelector(".branch");
			
			// can't go left of 0
			let buttonLeft = document.createElement("button");
			buttonLeft.onclick = () => 
			{
				data_parent_leaf.decrementCurrentLeafIndex();
				load_branched_thread();
			};
			buttonLeft.innerHTML = `<i class="fa-regular fa-angle-left"></i>`;
			branchSelectorElement.appendChild(buttonLeft);
			if((current_leaf_index <= 0))
			{
				buttonLeft.disabled = true;
			}
			
			let branchText = document.createElement("span");
			branchText.classList.add("flex-grow");
			branchText.classList.add("flex-shrink-0");
			branchText.innerHTML = `${current_leaf_index+1} / ${leaves_length}`;
			branchSelectorElement.appendChild(branchText);
			
			// can't go right of max 
			let buttonRight = document.createElement("button");
			buttonRight.onclick = () => 
			{
				data_parent_leaf.incrementCurrentLeafIndex();
				load_branched_thread();
			};
			buttonRight.innerHTML = `<i class="fa-regular fa-angle-right"></i>`;
			branchSelectorElement.appendChild(buttonRight);
			if((current_leaf_index >= leaves_length - 1))
			{
				buttonRight.disabled = true;
			}
		}
		
        if (human) {
            temp.querySelector(".text").innerHTML = `<p>${fake_convo[i]}</p>`
            main.appendChild(temp)
        }
        else{
            let clipboard = `<i class="fa-regular clipboard fa-clipboard"></i>`
            let copy_bar = `<div class="p-2 copy float-right">${clipboard} &nbsp; Copy code</div>`
            temp.querySelector(".text").innerHTML = fake_convo[i].replaceAll(bar, copy_bar).replaceAll(`<div class="p-4">`, "<div>") // fixes formatting for weird code divs
            main.appendChild(temp)
        }
	}
	
	// add buttons.
	copy_setup();
}

// Add the right event listeners for the little copy clipboard in code blocks.
function copy_setup() { // created by ChatGPT
    const clipboardBars = document.querySelectorAll('.copy');
    const codeElements = document.querySelectorAll('pre code');

// Add a click event listener to each clipboard bar
    clipboardBars.forEach((clipboardBar, index) => {
        clipboardBar.addEventListener('click', async () => {
            let clipboard = `<i class="fa-regular clipboard fa-clipboard"></i>`
            let copy_bar = `<div class="p-2 copy float-right">${clipboard} &nbsp; Copy code</div>`
            clipboardBar.innerHTML = `<icon class="fa-regular fa-check"></icon> &nbsp; Copied!`;
            setTimeout(() => {clipboardBar.outerHTML = copy_bar; copy_setup()}, 2000);
            // Get the code element corresponding to the clicked clipboard bar
            const codeElement = codeElements[index];

            // Get the text content of the code element
            const text = codeElement.textContent;

            // Copy the text to the clipboard
            await navigator.clipboard.writeText(text);
        });
    });

}

function getInnerText(className, propertyName) {
    // Get all elements with the given class name
    const elements = document.getElementsByClassName(className);

    // Initialize an empty array to store the inner text of each element
    const innerTextArray = [];

    // Iterate over the elements and add their inner text to the array
    for (let i = 0; i < elements.length; i++) {
        innerTextArray.push({ [propertyName]: elements[i].innerText });
    }

    // Return the array
    return innerTextArray;
}
function alternateValues(array1, array2) {
    return array1.map((val, i) => (i % 2 === 0) ? val : array2[i]);
}

// Open a new thread on a new instance of ChatGPT
function continue_thread(){
	let c = [];
	for (let i = 0; i < convo.length; i++) {
		let user = i % 2 === 0 ? "Me" : "ChatGPT";
		c.push({ [user]: htmlToPlainText(convo[i]) });
	}
    chrome.runtime.sendMessage({convo: c, type: 'b_continue_convo'});
}

document.querySelector("#continue").addEventListener("click", continue_thread);